instrumentation-api-semconv/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/net/PeerServiceResolver.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.instrumentation.api.instrumenter.net;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.Comparator;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+
+class PeerServiceResolver {
+
+  private static final Map<String, Map<ServiceMatcher, String>> MAPPING = new ConcurrentHashMap<>();
+
+  PeerServiceResolver(Map<String, String> peerServiceMapping) {
+    peerServiceMapping.forEach(
+        (key, serviceName) -> {
+          try {
+            String protocol = null;
+            URI uri;
+            if (key.contains("://")) {
+              uri = new URI(key);
+              protocol = uri.getScheme();
+            } else {
+              uri = new URI("http://" + key);
+            }
+            String host = uri.getHost();
+            Integer port = uri.getPort() == -1 ? null : uri.getPort();
+            Map<ServiceMatcher, String> matchers =
+                MAPPING.computeIfAbsent(host, x -> new ConcurrentHashMap<>());
+            matchers.putIfAbsent(new ServiceMatcher(protocol, port, uri.getPath()), serviceName);
+          } catch (URISyntaxException use) {
+            // TODO wire logging
+          }
+        });
+  }
+
+  boolean isEmpty() {
+    return MAPPING.isEmpty();
+  }
+
+  String resolveService(String protocol, String host, Integer port, String path) {
+    Map<ServiceMatcher, String> matchers = MAPPING.get(host);
+    if (matchers == null) {
+      return null;
+    }
+    return matchers.entrySet().stream()
+        .filter(entry -> entry.getKey().matches(protocol, port, path))
+        .max(Comparator.comparingInt(entry -> entry.getKey().getPathLength()))
+        .map(Map.Entry::getValue)
+        .orElse(null);
+  }
+
+  private static class ServiceMatcher {
+
+    private final String protocol;
+    private final Integer port;
+    private final String path;
```

Comment:
Null-safety annotations:
```suggestion
    @Nullable private final String protocol;
    @Nullable private final Integer port;
    @Nullable private final String path;
```